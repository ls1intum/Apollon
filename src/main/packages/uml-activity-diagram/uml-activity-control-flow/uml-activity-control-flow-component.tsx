import React, { FunctionComponent } from 'react';
import { Point } from '../../../utils/geometry/point';
import { UMLActivityControlFlow } from './uml-activity-control-flow';
import { ThemedPath, ThemedPolyline } from '../../../components/theme/themedComponents';

export const UMLActivityControlFlowComponent: FunctionComponent<Props> = ({ element }) => {
  let position = { x: 0, y: 0 };
  let direction: 'v' | 'h' = 'v';
  const path = element.path.map((point) => new Point(point.x, point.y));
  let distance =
    path.reduce(
      (length, point, i, points) => (i + 1 < points.length ? length + points[i + 1].subtract(point).length : length),
      0,
    ) / 2;

  for (let index = 0; index < path.length - 1; index++) {
    const vector = path[index + 1].subtract(path[index]);
    if (vector.length > distance) {
      const norm = vector.normalize();
      direction = Math.abs(norm.x) > Math.abs(norm.y) ? 'h' : 'v';
      position = path[index].add(norm.scale(distance));
      break;
    }
    distance -= vector.length;
  }

  const layoutText = (dir: 'v' | 'h') => {
    switch (dir) {
      case 'v':
        return {
          dx: 5,
          dominantBaseline: 'middle',
          textAnchor: 'start',
        };
      case 'h':
        return {
          dy: -5,
          dominantBaseline: 'text-after-edge',
          textAnchor: 'middle',
        };
    }
  };

  const fill = element.textColor ? { fill: element.textColor } : {};

  return (
    <g>
      <marker
        id={`marker-${element.id}`}
        viewBox="0 0 30 30"
        markerWidth="22"
        markerHeight="30"
        refX="30"
        refY="15"
        orient="auto"
        markerUnits="strokeWidth"
      >
        <ThemedPath d="M0,29 L30,15 L0,1" fillColor="none" strokeColor={element.strokeColor} />
      </marker>
      <ThemedPolyline
        points={element.path.map((point) => `${point.x} ${point.y}`).join(',')}
        strokeColor={element.strokeColor}
        fillColor="none"
        strokeWidth={1}
        markerEnd={`url(#marker-${element.id})`}
      />
      <text x={position.x} y={position.y} {...layoutText(direction)} pointerEvents="none" style={{ ...fill }}>
        {element.name}
      </text>
    </g>
  );
};

interface Props {
  element: UMLActivityControlFlow;
}
