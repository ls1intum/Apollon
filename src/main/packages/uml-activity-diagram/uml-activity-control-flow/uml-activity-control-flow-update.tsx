import React, { Component, ComponentClass } from 'react';
import { connect } from 'react-redux';
import { compose } from 'redux';
import styled from 'styled-components';
import { Button } from '../../../components/controls/button/button';
import { Divider } from '../../../components/controls/divider/divider';
import { ExchangeIcon } from '../../../components/controls/icon/exchange';
import { TrashIcon } from '../../../components/controls/icon/trash';
import { Textfield } from '../../../components/controls/textfield/textfield';
import { Header } from '../../../components/controls/typography/typography';
import { I18nContext } from '../../../components/i18n/i18n-context';
import { localized } from '../../../components/i18n/localized';
import { ModelState } from '../../../components/store/model-state';
import { UMLElementRepository } from '../../../services/uml-element/uml-element-repository';
import { UMLRelationshipRepository } from '../../../services/uml-relationship/uml-relationship-repository';
import { UMLActivityControlFlow } from './uml-activity-control-flow';
import { ColorButton } from '../../../components/controls/color-button/color-button';
import { StylePane } from '../../../components/style-pane/style-pane';

const Flex = styled.div`
  display: flex;
  align-items: baseline;
  justify-content: space-between;
`;

type State = { colorOpen: boolean };

class ActivityControlFlowUpdate extends Component<Props, State> {
  state = { colorOpen: false };

  private toggleColor = () => {
    this.setState((state) => ({
      colorOpen: !state.colorOpen,
    }));
  };

  render() {
    const { element } = this.props;

    return (
      <div>
        <section>
          <Flex>
            <Header gutter={false} style={{ flexGrow: 1 }}>
              {this.props.translate('packages.ActivityDiagram.ActivityControlFlow')}
            </Header>
            <ColorButton onClick={this.toggleColor} />
            <Button color="link" onClick={() => this.props.flip(element.id)}>
              <ExchangeIcon />
            </Button>
            <Button color="link" onClick={() => this.props.delete(element.id)}>
              <TrashIcon />
            </Button>
          </Flex>
          <Divider />
        </section>
        <section>
          <Textfield value={element.name} onChange={this.rename} autoFocus />
        </section>
        <StylePane
          open={this.state.colorOpen}
          element={element}
          onColorChange={this.props.update}
          lineColor
          textColor
        />
      </div>
    );
  }

  private rename = (name: string) => {
    this.props.update(this.props.element.id, { name });
  };
}

type OwnProps = {
  element: UMLActivityControlFlow;
};

type StateProps = {};

type DispatchProps = {
  update: typeof UMLElementRepository.update;
  delete: typeof UMLElementRepository.delete;
  flip: typeof UMLRelationshipRepository.flip;
};

type Props = OwnProps & StateProps & DispatchProps & I18nContext;

const enhance = compose<ComponentClass<OwnProps>>(
  localized,
  connect<StateProps, DispatchProps, OwnProps, ModelState>(null, {
    update: UMLElementRepository.update,
    delete: UMLElementRepository.delete,
    flip: UMLRelationshipRepository.flip,
  }),
);

export const UMLActivityControlFlowUpdate = enhance(ActivityControlFlowUpdate);
