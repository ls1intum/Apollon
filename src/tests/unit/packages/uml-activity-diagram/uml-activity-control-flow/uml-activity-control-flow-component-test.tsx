import * as React from 'react';
import { wrappedRender } from '../../../test-utils/render';
import { UMLActivityControlFlow } from '../../../../../main/packages/uml-activity-diagram/uml-activity-control-flow/uml-activity-control-flow';
import { UMLActivityControlFlowComponent } from '../../../../../main/packages/uml-activity-diagram/uml-activity-control-flow/uml-activity-control-flow-component';

it('render the uml-activity-control-flow-component', () => {
  const controlFlow: UMLActivityControlFlow = new UMLActivityControlFlow({
    id: 'd37b8ce3-17d2-4432-8fff-6c38ff2a1334',
    name: 'TestActivityComponent',
  });
  const { getByText, baseElement } = wrappedRender(
    <svg>
      <UMLActivityControlFlowComponent element={controlFlow} />
    </svg>,
  );
  expect(getByText(controlFlow.name)).toBeInTheDocument();
  expect(baseElement).toMatchSnapshot();
});
