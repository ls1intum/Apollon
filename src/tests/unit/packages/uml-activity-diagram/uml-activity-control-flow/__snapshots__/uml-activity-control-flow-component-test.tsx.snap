// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render the uml-activity-control-flow-component 1`] = `
<body>
  <div>
    <svg>
      <g>
        <marker
          id="marker-d37b8ce3-17d2-4432-8fff-6c38ff2a1334"
          markerHeight="30"
          markerUnits="strokeWidth"
          markerWidth="22"
          orient="auto"
          refX="30"
          refY="15"
          viewBox="0 0 30 30"
        >
          <path
            class="sc-gsnTZi iAPApG"
            d="M0,29 L30,15 L0,1"
            fill="none"
            stroke="black"
          />
        </marker>
        <polyline
          class="sc-bczRLJ gQkeMM"
          fill="none"
          marker-end="url(#marker-d37b8ce3-17d2-4432-8fff-6c38ff2a1334)"
          points="0 0,200 100"
          stroke="black"
          stroke-width="1"
        />
        <text
          dominant-baseline="text-after-edge"
          dy="-5"
          pointer-events="none"
          text-anchor="middle"
          x="100"
          y="50"
        >
          TestActivityComponent
        </text>
      </g>
    </svg>
  </div>
</body>
`;
